File: mobile/src/types/expense.ts
export interface Category {
  id: string;
  user_id?: string | null;
  name: string;
  icon?: string | null;
  color?: string | null;
  is_system: boolean;
  type: 'expense' | 'income' | 'bike';
  created_at: string;
}

export interface Expense {
  id: string;
  user_id: string;
  category_id: string;
  amount: number | string;
  currency: string;
  description: string;
  merchant?: string | null;
  payment_method: string;
  expense_date: string;
  notes?: string | null;
  idempotency_key?: string | null;
  created_at: string;
  updated_at: string;
  category?: Category | null;
}

export interface ExpenseCreate {
  amount: number;
  currency?: string;
  category_id: string;
  description: string;
  merchant?: string;
  payment_method?: string;
  expense_date: string;
  notes?: string;
  idempotency_key?: string;
}

export interface ExpenseUpdate {
  amount?: number;
  currency?: string;
  category_id?: string;
  description?: string;
  merchant?: string;
  payment_method?: string;
  expense_date?: string;
  notes?: string;
}

export interface CategorySpend {
  category_id: string;
  category_name: string;
  category_icon?: string | null;
  category_color?: string | null;
  total_amount: number | string;
  percentage: number;
}

export interface ExpenseSummary {
  today_total: number | string;
  this_month_total: number | string;
  total_expenses_count: number;
  category_breakdown: CategorySpend[];
}

export interface PaginatedExpenses {
  data: Expense[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface Income {
  id: string;
  user_id: string;
  amount: number | string;
  currency: string;
  source: string;
  description?: string | null;
  income_date: string;
  is_recurring: boolean;
  recurrence_period?: string | null;
  created_at: string;
  updated_at: string;
}

export interface IncomeCreate {
  amount: number;
  currency?: string;
  source: string;
  description?: string;
  income_date: string;
  is_recurring?: boolean;
  recurrence_period?: string;
}

File: backend/app/schemas/expense.py
from datetime import date, datetime
from decimal import Decimal
from uuid import UUID
from typing import Optional, List
from pydantic import BaseModel, ConfigDict, Field
from app.schemas.category import CategoryResponse

class ExpenseBase(BaseModel):
    amount: Decimal = Field(..., gt=0, decimal_places=2, max_digits=12)
    currency: str = Field("INR", min_length=3, max_length=3)
    category_id: UUID
    description: str = Field(..., min_length=1, max_length=255)
    merchant: Optional[str] = Field(None, max_length=255)
    payment_method: str = Field("UPI", max_length=50)
    expense_date: date
    notes: Optional[str] = None

class ExpenseCreate(ExpenseBase):
    idempotency_key: Optional[UUID] = None

class ExpenseUpdate(BaseModel):
    amount: Optional[Decimal] = Field(None, gt=0, decimal_places=2, max_digits=12)
    currency: Optional[str] = Field(None, min_length=3, max_length=3)
    category_id: Optional[UUID] = None
    description: Optional[str] = Field(None, min_length=1, max_length=255)
    merchant: Optional[str] = Field(None, max_length=255)
    payment_method: Optional[str] = Field(None, max_length=50)
    expense_date: Optional[date] = None
    notes: Optional[str] = None

class ExpenseBulkCreate(BaseModel):
    expenses: List[ExpenseCreate]

class ExpenseResponse(ExpenseBase):
    id: UUID
    user_id: UUID
    idempotency_key: Optional[UUID] = None
    created_at: datetime
    updated_at: datetime
    category: Optional[CategoryResponse] = None

    model_config = ConfigDict(from_attributes=True)

class CategorySpend(BaseModel):
    category_id: UUID
    category_name: str
    category_icon: Optional[str] = None
    category_color: Optional[str] = None
    total_amount: Decimal
    percentage: float

class ExpenseSummaryResponse(BaseModel):
    today_total: Decimal
    this_month_total: Decimal
    total_expenses_count: int
    category_breakdown: List[CategorySpend]
